Seminar130919/msk_inf_111540.cpp

```diff
@@ -0,0 +1,156 @@
+#include <iostream>
+#include <vector>
+#include <queue>
+#include <algorithm>
+
+/* Дан неориентированный невзвешенный граф. Необходимо посчитать количество его компонент связности и вывести их. */
+
+class Graph {
+public:
+    typedef size_t Vertex;
+
+    Graph(size_t _vertex_count, bool _is_directed) {
+        is_directed = _is_directed;
+        vertex_count = _vertex_count;
+        edges_count = 0;
+    }
+
+    bool directed() {
+        return is_directed;
+    }
+
+    virtual size_t get_vertex_deg(const Vertex &v) const = 0;
+
+    size_t get_vertex_count() const {
+        return vertex_count;
+    }
+
+    size_t get_edges_count() const {
+        return edges_count;
+    }
+
+    virtual void add_edge(const Vertex &start, const Vertex &finish) {
+        edges_count++;
+    }
+
+    virtual std::vector<Vertex> get_all_neighbours(const Vertex &v) const = 0;
+
+protected:
+    size_t vertex_count, edges_count;
+    bool is_directed;
+};
+
+class AdjListGraph : public Graph {
+public:
+    AdjListGraph(size_t _vertex_count, bool _is_directed) : Graph(_vertex_count, _is_directed) {
+        Vertexes.resize(vertex_count);
+    }
+
+    AdjListGraph(bool _is_directed, const std::vector<std::vector<int>> &matrix) : Graph(
+            matrix.size(), _is_directed) {
+        Vertexes.resize(vertex_count);
+        read_matrix(matrix);
+    }
+
+    virtual void add_edge(const Vertex &start, const Vertex &finish) override {
+        Graph::add_edge(start, finish);
+        Vertexes[start].push_back(finish);
+        if (!is_directed) {
+            Vertexes[finish].push_back(start);
+        }
+    }
+
+    virtual size_t get_vertex_deg(const Vertex &v) const override {
+        return Vertexes[v].size();
+    }
+
+
+    virtual std::vector<Vertex> get_all_neighbours(const Vertex &v) const override {
+        return Vertexes[v];
+    }
+
+private:
+    std::vector<std::vector<Vertex>> Vertexes;
+
+    void read_matrix(const std::vector<std::vector<int>> &matrix) {
+        if (matrix.size() != get_vertex_count()) {
+            throw std::out_of_range("wrong matrix size");
+        }
+        for (Vertex i = 0; i < matrix.size(); i++) {
+            for (Vertex j = 0; j < matrix[i].size(); j++) {
+                if (matrix[i][j] != 0) {
+                    if (j >= i || is_directed) {
+                        add_edge(i, j);
+                    }
+                }
+            }
+        }
+    };
+};
+
+class WeightedAgjListGraph : public AdjListGraph {
+private:
+    std::vector<int> weight;
+public:
+
+};
+namespace GraphAlgorithms {
+    enum VertexColours {
```

Comment:
Enumerators (for both scoped and unscoped enums) should be named either like constants or like macros: either kEnumName or ENUM_NAME.